fili-core/src/main/java/com/yahoo/bard/webservice/web/handlers/CacheWeightCheckRequestHandler.java

```diff
@@ -0,0 +1,225 @@
+            QueryWeightUtil queryWeightUtil,
+            ObjectMapper mapper,
+            @NotNull DataCache<?> dataCache,
+            QuerySigningService<?> querySigningService
+    ) {
+        super(
+                next,
+                webService,
+                queryWeightUtil,
+                mapper
+        );
+        this.dataCache = (TupleDataCache<String, Long, String>) dataCache;
+        this.querySigningService = (QuerySigningService<Long>) querySigningService;
+    }
+
+    @Override
+    public boolean handleRequest(
+            final RequestContext context,
+            final DataApiRequest request,
+            final DruidAggregationQuery<?> druidQuery,
+            final ResponseProcessor response
+    ) {
+        // Heuristic test to let requests with very low estimated cardinality directly through
+        if (queryWeightUtil.skipWeightCheckQuery(druidQuery)) {
+            return next.handleRequest(context, request, druidQuery, response);
+        }
+
+        BardQueryInfo.getBardQueryInfo().incrementCountWeightCheck();
+        final WeightCheckResponseProcessor weightCheckResponse =
+                new WeightCheckResponseProcessor(response);
+        final DruidAggregationQuery<?> weightEvaluationQuery = queryWeightUtil.makeWeightEvaluationQuery(druidQuery);
+        Granularity granularity = druidQuery.getInnermostQuery().getGranularity();
+        final long queryRowLimit = queryWeightUtil.getQueryWeightThreshold(granularity);
+
+        try {
+            LOG.debug("Weight query {}", writer.writeValueAsString(weightEvaluationQuery));
+        } catch (JsonProcessingException e) {
+            LOG.warn("Weight Query json exception:", e);
+        }
+
+        CacheService cacheService = new CacheService();
+        String cacheKey = null;
+        try {
+            cacheKey = getKey(druidQuery, mapper, writer);
+        } catch (JsonProcessingException e) {
+            LOG.warn("Cache key cannot be built: ", e);
+        }
+
+        WeightCheckRequestHandler weightCheckRequestHandler =
+                new WeightCheckRequestHandler(next, webService, queryWeightUtil, mapper);
+        SuccessCallback weightQuerySuccess = null;
+        // Construct success callback based on cache hit or cache miss
+        if (context.isReadCache()) {
+            String cacheResponse =
+                    cacheService.readCache(context, dataCache, querySigningService, druidQuery, cacheKey);
+            // Cache hit
+            if (cacheResponse != null) {
+                weightQuerySuccess = buildCacheSuccessCallback(
+                        context,
+                        request,
+                        druidQuery,
+                        weightCheckResponse,
+                        queryRowLimit,
+                        cacheKey,
+                        cacheService,
+                        weightCheckRequestHandler,
+                        false
+                );
+            }
+        } else {
```

Comment:
With the simplified call stack and prebuilding the callbacks, this can look like: 
```suggestion
       try {
            String cacheResponse = cacheService.readCache(context, druidQuery);
            // There is a cache hit, so no more cache interactions are necessary.

            if (cacheResponse != null) {
                JsonNode rootNode = mapper.readTree(cacheResponse);
                classicCallback.invoke(rootNode);
                return true;
            }
        } catch (JsonMappingException e) {
            e.printStackTrace();
        } catch (JsonProcessingException e) {
            e.printStackTrace();
        }```

Just directly invoke the callback, no request handler delegation.  The success callback will handle that.